equinox/nn/_weight_norm.py

```diff
@@ -0,0 +1,104 @@
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+from beartype.typing import Callable, Generic, Optional, TypeVar
+from jaxtyping import Array, PRNGKeyArray, PyTree
+
+from .._module import field, Module
+from .._tree import tree_at
+
+
+_Layer = TypeVar("_Layer")
+
+
+def _norm_except_axis(
+    v: Array, norm: Callable[[PyTree[Array]], Array], axis: Optional[int]
+):
+    if axis is None:
+        return norm(v)
+    else:
+        return jax.vmap(norm, in_axes=axis, out_axes=axis)(v)
+
+
+class WeightNorm(Module, Generic[_Layer]):
+    r"""
+    Applies weight normalisation to a given parameter.
+
+    Given a weight matrix $\mathbf{W}$, computes the follow reparametrization:
+
+    $\mathbf{W} = g \frac{\mathbf{v}}{\lVert \mathbf{v} \rVert}$
+
+    where $g$ is initially chosen to equal $\lVert \mathbf{v} \rVert$
+    , and $\mathbf{v}$ is initially chosen as $\mathbf{W}$ .
+
+    ??? cite
+        [Weight Normalisation](https://arxiv.org/abs/1602.07868)
+
+        ```bibtex
+        @article{DBLP:journals/corr/SalimansK16,
+        author       = {Tim Salimans and
+                        Diederik P. Kingma},
+        title        = {Weight Normalisation: {A} Simple
+                        Reparameterization to Accelerate
+                        Training of Deep Neural Networks},
+        journal      = {CoRR},
+        volume       = {abs/1602.07868},
+        year         = {2016},
+        url          = {http://arxiv.org/abs/1602.07868},
+        eprinttype   = {arXiv},
+        eprint       = {1602.07868},
+        timestamp    = {Mon, 13 Aug 2018 16:47:07 +0200},
+        biburl       = {https://dblp.org/rec/journals/corr/SalimansK16.bib},
+        bibsource    = {dblp computer science bibliography, https://dblp.org}
+        }
+        ```
+    """
+
+    layer: _Layer
+    v: Array
+    g: Array
+    weight_name: str = field(static=True)
+    axis: Optional[int] = field(static=True)
+    _norm: Callable[[Array], Array]
+
+    def __init__(
+        self,
+        layer: _Layer,
+        weight_name: str = "weight",
+        axis: Optional[int] = 0,
+    ):
+        """**Arguments:**
+
+        - `layer`: The layer to wrap. Usually a [`equinox.nn.Linear`][] or
+        a convolutional layer (e.g. [`equinox.nn.Conv2d`][]).
+        - `weight_name`: The name of the layer's parameter (a JAX array) to apply
+        weight normalisation to.
+        - `axis`: The norm is computed across every axis except this one.
+        If `None`, compute across every axis.
+        """
+        self.layer = layer
+        self.weight_name = weight_name
+        self.axis = axis
+
+        self.v = getattr(layer, weight_name)
```

Comment:
Oh, I've just spotted a minor bug in this implementation. We now have both `self.v` and `self.layer.weight`, both of which are the "weight" of this layer.

[This means each version of the weight is treated as a separate copy.](https://docs.kidger.site/equinox/faq/#how-to-share-a-layer-across-two-parts-of-a-model)

I think the fix is to remove `self.v` from `__init__` entirely. And then in the forward pass, do:
```python
v = getattr(self.layer, self.weight_name)
weight = self.g * v / self._norm(v)
layer = tree_at(lambda l: getattr(l, self.weight_name), self.layer, weight)
return layer(x)
```

Reply:
Yeah, I initially chose to implement it this way to cleanly separate what is actually being trained - `v` and `g` and not the `weight` of the layer directly. But that distinction is definitely not worth wasting memory.
This makes me think though, maybe it'd be worth it to add a "commit" function or something like that, which returns the contained layer with the weight matrix set to its final value by evaluating `g * v/norm(v)`. This can be done manually by the user also of course, but that's not very ergonomic. WDYT?

Reply:
So something like this:
```python
    def commit(self) -> PyTree:
        v = getattr(self.layer, self.weight_name)
        weight = v * self.g / self._norm(v)
        layer = tree_at(lambda l: getattr(l, self.weight_name), self.layer, weight)
        return layer
```

Which would simplify forward like so:
```python
        return self.commit()(x)
```.

Or alternatively we'd change the way this behaves during inference, where it's not necessary to recalculate `g * v/norm(v)` every forward pass - where we could overwrite `self.layer.weight` with the final weight matrix and simply reverse the operation for training mode.

Reply:
No strong feelings on whether to pull this functionality out into a `commit` method. Do it or don't as you prefer I think! :)

As for inference-time optimisations, this is an interesting point. I think I'd say for now that we shouldn't worry about this: there's a lot of these optimisations that we don't do at the moment, and they often tend to also involve fusing layers together, this kind of thing. So perhaps best to find another way to tackle this problem.

Reply:
I'll leave it be for now then - if someone creates an issue for this, I'll be happy to add it. Feel free to tag me if this should happen!

Pushed all changes now :)